Add scaladoc docs (#2012)
scala/docs.scala-lang: scala3/scaladoc.md

---
layout: singlepage-overview

> title: New features for Scaladoc

Author: Could you elaborate what headers should I use? I was basing on this [page](https://github.com/scala/docs.scala-lang/blob/main/scala3/new-in-scala3.md) while choosing header settings

> title: New features for Scaladoc

Author: Page shoud be visible now from both root page in `Scala 3` section or from `/scala3` suburi
![obraz](https://user-images.githubusercontent.com/32793002/117834210-439ac180-b277-11eb-8105-b15810938e1e.png)

> title: New features for Scaladoc

Reviewer: @BarkingBad  @julienrf  What do you think of changing the title of the document to: "All new Scaladoc for Scala 3"? Just to emphasize its relation to Scala 3 again...

> title: New features for Scaladoc

Author: Sound great, will change :D

---

New Scala version brings us completely new Scaladoc rewritten from scratch. 

> In this article you can find highlights of new features that are or will be introduced to Scaladoc.

Author: Do you think two h1 chapters would be sufficient?

> In this article you can find highlights of new features that are or will be introduced to Scaladoc.

Reviewer: Sorry for nitpicking: I think typically we only have one `h1` per page, but I might be wrong. So maybe you can move all headlines just one level? h1 -> h2, h2 -> h3, ...

> In this article you can find highlights of new features that are or will be introduced to Scaladoc.

Author: I can do that

For general reference, visit [Scaladoc manual](https://dotty.epfl.ch/docs/usage/scaladoc/)

## Markdown syntax

The biggest change introduced in new Scaladoc is change of default language for styling docstring. So far it has been Wikidoc syntax.
New Scaladoc can parse legacy `Wikidoc` syntax, however Markdown has been chosen for primary language for formatting comments.
To switch back to `Wikidoc` one can pass global flag before running `doc` task or can define it for specific comment via `@syntax wiki` directive.

For more information how to use full power of docstings, check out [Scaladoc docstrings][scaladoc-docstrings]

## Static site

Scaladoc brings easy way for creating static sites for both documentation and blog posts in the similar way as Jekyll does.
Thanks to that feature you can store your documentation along-side with Scaladoc API docs output in a very convenient way.

For more info about how to configure static site documentation check out [Static documentation][static-documentation] chapter

![](../resources/images/scala3/scaladoc/static-site.png)

## Blog posts

> Blog posts are subtype of static sites. To find more information about them look at dedicated page for [Built-in blog][[built-in-blog]]

Reviewer 2: ```suggestion
Blog posts are a specific type of static site. To find more information about them look at the dedicated page for [Built-in blog][built-in-blog]
```

Author: Will fix, thnx

![](../resources/images/scala3/scaladoc/blog-post.png)

## Social links

> Scaladoc provide easy way to configure your social media links e. g. twitter or gitter. For more information look at subsection of 

Reviewer 2: ```suggestion
Scaladoc provides an easy way to configure your social media links e. g. twitter or gitter. For more information look at the subsection of 
```

Author: Will fix, thnx

[Scaladoc settings][social-links] chapter.

![](../resources/images/scala3/scaladoc/social-links.png)

## Snippets compiler

One of the experimental features of Scaladoc will be snippets compiler. This tool will allow you to compile snippets that you attach to your docstring 

> to check whether that they actually behave as intended, e. g. compile or throw some exception. The feature is very similar to `tut` or `mdoc` tools,

Reviewer 2: ```suggestion
to check that they actually behave as intended, e. g. compile or throw some exception. The feature is very similar to `tut` or `mdoc` tools,
```

Author: Will fix, thnx

---
layout: singlepage-overview
title: New features for Scaladoc
---

> New Scala version brings us completely new Scaladoc rewritten from scratch. 

Reviewer: ```suggestion
The new Scala version 3 comes with a completely new implementation of the documentation generator _Scaladoc_, rewritten from scratch. 
```

Author: Fixed

In this article you can find highlights of new features that are or will be introduced to Scaladoc.
For general reference, visit [Scaladoc manual](https://dotty.epfl.ch/docs/usage/scaladoc/)

# New features

## Markdown syntax

> The biggest change introduced in new Scaladoc is change of default language for styling docstring. So far it has been Wikidoc syntax.
> New Scaladoc can parse legacy `Wikidoc` syntax, however Markdown has been chosen for primary language for formatting comments.
> To switch back to `Wikidoc` one can pass global flag before running `doc` task or can define it for specific comment via `@syntax wiki` directive.
> 
> For more information how to use full power of docstings, check out [Scaladoc docstrings][scaladoc-docstrings]

Reviewer: ```suggestion
The biggest change introduced in the new version of Scaladoc is the change of the default language for docstrings. So far Scaladoc only supported Wikidoc syntax.
The new Scaladoc can still parse legacy `Wikidoc` syntax, however Markdown has been chosen as a primary language for formatting comments.
To switch back to `Wikidoc` one can pass a global flag before running the `doc` task or one can define it for specific comments via the `@syntax wiki` directive.

For more information on how to use the full power of docstings, check out [Scaladoc docstrings][scaladoc-docstrings]
```

Author: Fixed

## Static site

> Scaladoc brings easy way for creating static sites for both documentation and blog posts in the similar way as Jekyll does.
> Thanks to that feature you can store your documentation along-side with Scaladoc API docs output in a very convenient way.
> 
> For more info about how to configure static site documentation check out [Static documentation][static-documentation] chapter

Reviewer: ```suggestion
Scaladoc also provides an easy way for creating **static sites** for both documentation and blog posts in the similar way as Jekyll does.
Thanks to this feature, you can store your documentation along-side with the generated Scaladoc API in a very convenient way.

For more information on how to configure the generation of static sites check out [Static documentation][static-documentation] chapter
```

Author: Fixed

![](../resources/images/scala3/scaladoc/static-site.png)

## Blog posts

> Blog posts are a specific type of static site. To find more information about them look at the dedicated page for [Built-in blog][built-in-blog]
> 
> ![](../resources/images/scala3/scaladoc/blog-post.png)

Reviewer: ```suggestion
Blog posts are a specific type of static sites. In the Scaladoc manual you can find additional information about how to work with [blog posts][built-in-blog].

![](../resources/images/scala3/scaladoc/blog-post.png)
```

Author: Fixed

## Social links

> Scaladoc provides an easy way to configure your social media links e. g. twitter or gitter. For more information look at the subsection of 
> [Scaladoc settings][social-links] chapter.
> 
> ![](../resources/images/scala3/scaladoc/social-links.png){: style="width: 180px"}

Reviewer: ```suggestion
Furthermore, Scaladoc provides an easy way to configure your [social media links][social-links] e.g. Twitter or Gitter.

![](../resources/images/scala3/scaladoc/social-links.png){: style="width: 180px"}
```

Author: Fixed

# Experimental features

> These features are not stable to be released with scaladoc, however we are happy to hear your feedback. Each feature has its own thread at scala-lang contributors site, where you can share your opinions.

Reviewer: ```suggestion
The following features are not stable to be released with scaladoc, however we are happy to hear your feedback. Each feature has its own thread at scala-lang contributors site, where you can share your opinions.
```

Author: Fixed

> ## Snippets compiler
> 
> One of the experimental features of Scaladoc will be snippets compiler. This tool will allow you to compile snippets that you attach to your docstring 
> to check that they actually behave as intended, e. g. compile or throw some exception. The feature is very similar to `tut` or `mdoc` tools, 
> but will be shipped with Scaladoc out of the box for easy setup. Making snippets interactive, e. g. user could edit them and compile in the browser is taken into account, though it is not in scope yet.
> 
> For more information you can follow this [thread](https://contributors.scala-lang.org/t/snippet-validation-in-scaladoc-for-scala-3/4976)
> 
> ![](../resources/images/scala3/scaladoc/snippet-compiler2.gif)
> ![](../resources/images/scala3/scaladoc/snippet-compiler1.gif)

Reviewer: ```suggestion
## Snippets compiler

One of the experimental features of Scaladoc will be a snippets compiler. This tool will allow you to compile snippets that you attach to your docstring 
to check that they actually behave as intended, e. g. compile or throw some exception. The feature is very similar to `tut` or `mdoc` tools, 
but will be shipped with Scaladoc out of the box for easy setup and integration into your project. Making snippets interactive, e. g. user could edit them and compile in the browser is under consideration, though it is not in scope yet.

For more information you can follow this [thread](https://contributors.scala-lang.org/t/snippet-validation-in-scaladoc-for-scala-3/4976)

![](../resources/images/scala3/scaladoc/snippet-compiler2.gif)
![](../resources/images/scala3/scaladoc/snippet-compiler1.gif)
```

Author: Fixed

## Type-signature browsing

> Haskell programmers are probably familiar with Hoogle - documentation search engine that allows you browse the docs by signatures rather than symbolic 
> names. Since many Scala developers are also functional programming fans, we decided to add similar functionality for the Scaladoc. The whole tool is based on [Inkuire](https://github.com/VirtusLab/Inkuire) which works with Kotlin and is being currently adapted to fit Scala3 type system. 
> 
> ![](../resources/images/scala3/scaladoc/inkuire.gif)
> 
> To be up-to-date with this feature you can follow this [thread](https://contributors.scala-lang.org/t/pre-sip-scaladoc-search-by-type-signature/1604/15).
> Note that this is very basic demo of this feature. We still don't support type lambdas, higher-kinded types and many more scala features.

Reviewer: ```suggestion
Haskell programmers are probably familiar with Hoogle - a documentation search engine that allows you browse the docs by signatures rather than symbolic 
names. Since many Scala developers are also functional programming fans, we decided to add a similar functionality to Scaladoc. The whole tool is based on [Inkuire](https://github.com/VirtusLab/Inkuire), which works with Kotlin and is being currently adapted to Scala 3. 

![](../resources/images/scala3/scaladoc/inkuire.gif)

To be up-to-date with this feature you can follow this [thread](https://contributors.scala-lang.org/t/pre-sip-scaladoc-search-by-type-signature/1604/15).
Note that this is very basic demo of this feature. We still don't support type lambdas, higher-kinded types, and many more Scala features.
```

Author: Fixed

> Note that this is very basic demo of this feature. We still don't support type lambdas, higher-kinded types and many more scala features.

Reviewer 3: ```suggestion
Note that this feature is in an early stage of development. We still don't support type lambdas, higher-kinded types, and some other Scala features.
```

Author: Fixed
